### PR TITLE
CB-6681 Handle cases when [only CM/only CDH/ neither] changes properly

### PR DESCRIPTION
**Pull request by schfeca75**
@keyki or @pdarvasi pls review

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java [hunk 1]
```diff
@@ -32,26 +35,52 @@
...
+        }
+        return clusterManagerUpdateNeeded;
     }
 
-    public void upgradeCluster(long stackId, Image image) {
-        String clusterManagerVersion = image.getVersion();
+    public boolean upgradeCluster(long stackId, StatedImage currentImage, StatedImage targetImage) {
+        Image targetIm = targetImage.getImage();
+        Image currentIm = currentImage.getImage();
+        String currentRuntimeBuildNumber = currentIm.getStackDetails().getStackBuildNumber();
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java:54]: Pls. ensure this is NPE safe

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java [hunk 2]
```diff
@@ -32,26 +35,52 @@
...
+        boolean clusterManagerUpdateNeeded = !currentIm.getCmBuildNumber().equals(targetIm.getCmBuildNumber());
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java:55]: Pls. ensure this is NPE safe

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java [hunk 3]
```diff
@@ -32,26 +35,52 @@
...
+        boolean clusterRuntimeUpgradeNeeded = !currentRuntimeBuildNumber.equals(targetIm.getStackDetails().getStackBuildNumber());
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java:56]: Pls. ensure this is NPE safe

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java [hunk 4]
```diff
@@ -32,26 +35,52 @@
...
 
-        flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_MANAGER_UPGRADE_FINISHED, clusterManagerVersion);
+        if (clusterManagerUpdateNeeded) {
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java:58]: could you pls add some extra logging here for the different clauses?

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java [hunk 5]
```diff
@@ -32,26 +35,52 @@
     @Inject
     private StackUpdater stackUpdater;
 
-    public void upgradeClusterManager(long stackId, StatedImage targetImage) {
+    public boolean upgradeClusterManager(long stackId, StatedImage currentImage, StatedImage targetImage) {
+        String currentCmBuildNumber = currentImage.getImage().getCmBuildNumber();
+        boolean clusterManagerUpdateNeeded = !currentCmBuildNumber.equals(targetImage.getImage().getCmBuildNumber());
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java:40]: clusterManagerUpdateNeeded and clusterRuntimeUpgradeNeeded are used twice, could you pls extract those to separate methods?

File: core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeServiceTest.java [hunk 6]
```diff
@@ -43,6 +78,124 @@ public void setUp() {
         MockitoAnnotations.initMocks(this);
     }
 
+    @Test
```
Comment (pdarvasi) [inline core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeServiceTest.java:81]: Nice coverage, well done!

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java (hunk 4 above)
Comment (schfeca75) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java:58]: The flowMessageService.fireEventAndLog() method will log the different kind of messages. I think it should be enough. What do you think?

File: core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeServiceTest.java (hunk 6 above)
Comment (schfeca75) [reply core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeServiceTest.java:86]: thank you!

**Pull request merged**